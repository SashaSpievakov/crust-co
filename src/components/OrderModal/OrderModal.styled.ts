import styled from 'styled-components';
import { BsXLg } from 'react-icons/bs';

import { ButtonPay } from 'src/styles/Buttons.styled';

const Aricle = styled.article`
  position: fixed;
  top: 50%;
  left: 50%;
  width: 35rem;
  height: 47rem;
  text-align: center;
  background: ${({ theme }) => theme.colors.fourth};
  transform: translate(-50%, -50%);
  z-index: 10;
  border-radius: 15px;
  box-shadow: 0 5px 20px 0 rgba(0, 0, 0, 0.05);

  @media screen and (max-width: 720px) {
    width: 30rem;
    height: 40rem;
  }

  @media screen and (max-width: 550px) {
    width: 25rem;
    height: 38rem;
  }

  @media screen and (max-width: 420px) {
    width: 18rem;
    height: 36rem;
  }
`;

const Header = styled.h3`
  font-size: 2.5rem;
  margin-top: 1.5rem;

  @media screen and (max-width: 720px) {
    font-size: 2rem;
  }

  @media screen and (max-width: 550px) {
    font-size: 1.7rem;
    margin-top: 1rem;
  }
`;

const SubHeader = styled.h4`
  font-size: 1.3rem;
  font-weight: 400;
  width: 70%;
  margin: 0 auto 3%;

  @media screen and (max-width: 720px) {
    font-size: 1.15rem;
  }

  @media screen and (max-width: 550px) {
    width: 80%;
  }

  @media screen and (max-width: 420px) {
    width: 85%;
    font-size: 1rem;
  }
`;

const Cross = styled(BsXLg)`
  position: absolute;
  top: 15px;
  right: 15px;
  font-size: 25px;
  opacity: 0.3;
  border-radius: 5px;
  cursor: pointer;

  &:hover {
    opacity: 0.7;
  }

  &:focus-visible {
    outline: 3px solid ${({ theme }) => theme.colors.third};
    outline-offset: 3px;
  }

  @media screen and (max-width: 550px) {
    font-size: 20px;
  }
`;

const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  row-gap: 2em;
  width: 70%;
  margin: 0 auto;

  @media screen and (max-width: 720px) {
    row-gap: 1.5em;
  }

  @media screen and (max-width: 550px) {
    width: 80%;
    row-gap: 1.3em;
  }

  @media screen and (max-width: 420px) {
    width: 85%;
    row-gap: 1.1em;
  }
`;

const Group = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
  align-items: start;
`;

const Label = styled.label`
  font-size: 1.3em;
  font-weight: 700;

  @media screen and (max-width: 720px) {
    font-size: 1.15rem;
  }

  @media screen and (max-width: 420px) {
    font-size: 1rem;
  }
`;

const Input = styled.input`
  font-size: 1.2em;
  width: 100%;
  padding: 0.5em 0.7em;
  color: ${({ theme }) => theme.colors.third};
  background-color: ${({ theme }) => theme.colors.gray};
  border: 3px solid ${({ theme }) => theme.colors.secondaryTransparent};
  border-radius: 10px;

  &:focus {
    border: 3px solid ${({ theme }) => theme.colors.secondary};
  }

  @media screen and (max-width: 720px) {
    padding: 0.3em 0.5em;
  }
`;

const ModalButton = styled(ButtonPay)`
  margin-top: 1rem;

  @media screen and (max-width: 550px) {
    padding: 0.8rem 1rem;
  }
`;

export {
  Aricle,
  Header,
  SubHeader,
  Cross,
  Form,
  Group,
  Label,
  Input,
  ModalButton,
};
