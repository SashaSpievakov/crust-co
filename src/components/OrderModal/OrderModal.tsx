import { KeyboardEvent, useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { ModalBg } from 'src/styles/Base.styled';

import SuccessModal from '../SuccessModal/SuccessModal';
import {
  Aricle,
  Cross,
  Form,
  Group,
  Header,
  Input,
  Label,
  ModalButton,
  SubHeader,
} from './OrderModal.styled';

interface OrderModalProps {
  setIsOpen: (value: boolean) => void;
}

interface FormValues {
  name: string;
  phone: string;
  city: string;
  address?: string;
}

const OrderModal = ({ setIsOpen }: OrderModalProps) => {
  const [isSubmitted, setIsSubmitted] = useState(false);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({ mode: 'onChange' });

  const onSubmit: SubmitHandler<FormValues> = () => {
    setIsSubmitted(true);
  };

  const handlePhoneInputKeyDownClick = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'e' || e.key === '-') {
      e.preventDefault();
    }
  };

  const handleCrossKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') setIsOpen(false);
  };

  return !isSubmitted ? (
    <>
      <ModalBg onClick={() => setIsOpen(false)} />
      <Aricle>
        <Header>Order</Header>
        <SubHeader>
          Please provide your information, and our operator will get in touch
          with you as soon as possible to finish your order
        </SubHeader>
        <Cross
          onClick={() => setIsOpen(false)}
          onKeyDown={(e) => handleCrossKeyDownClick(e)}
          tabIndex={0}
          data-testid="modalCross"
        />
        <Form onSubmit={handleSubmit(onSubmit)}>
          <Group>
            <Label htmlFor="inputName">Name</Label>
            <Input id="inputName" {...register('name', { required: true })} />
            {errors.name && <span>This field is required</span>}
          </Group>

          <Group>
            <Label htmlFor="inputPhone">Phone Number</Label>
            <Input
              id="inputPhone"
              type="number"
              onKeyPress={(e) => handlePhoneInputKeyDownClick(e)}
              {...register('phone', {
                required: {
                  value: true,
                  message: 'This field is required',
                },
                maxLength: {
                  value: 10,
                  message: 'The phone number should have 10 characters',
                },
                minLength: {
                  value: 10,
                  message: 'The phone number should have 10 characters',
                },
              })}
            />
            {errors.phone && <span>{errors.phone.message}</span>}
          </Group>

          <Group>
            <Label htmlFor="inputCity">City</Label>
            <Input id="inputCity" {...register('city', { required: true })} />
            {errors.city && <span>This field is required</span>}
          </Group>

          <Group>
            <Label htmlFor="inputAdddress">Adddress (optional)</Label>
            <Input id="inputAdddress" {...register('address')} />
          </Group>

          <ModalButton type="submit">
            <span>Send</span>
          </ModalButton>
        </Form>
      </Aricle>
    </>
  ) : (
    <SuccessModal setIsOpen={setIsOpen} />
  );
};
export default OrderModal;
