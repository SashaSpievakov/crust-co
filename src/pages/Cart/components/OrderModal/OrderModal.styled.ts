import { ButtonMedium, ButtonSecondary } from 'src/styles/Buttons.styled';
import styled from 'styled-components';

const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  row-gap: 2em;
  margin: 0 80px;
  width: 400px;
  margin: 0 auto;

  @media screen and (max-width: 720px) {
    row-gap: 1.5em;
    width: 350px;
    margin: 0 30px;
  }

  @media screen and (max-width: 550px) {
    row-gap: 1.3em;
  }

  @media screen and (max-width: 420px) {
    width: 260px;
    margin: 0 auto;
    row-gap: 1.1em;
  }
`;

const ModalButton = styled(ButtonMedium)`
  margin-top: 1rem;

  @media screen and (max-width: 550px) {
    padding: 0.8rem 0.9rem;
  }

  @media screen and (max-width: 420px) {
    padding: 0.6rem 0.7rem;
  }
`;

const CloseButton = styled(ButtonSecondary)`
  margin-top: 5%;

  @media screen and (max-width: 550px) {
    padding: 0.8rem 0.9rem;
  }

  @media screen and (max-width: 420px) {
    padding: 0.6rem 0.7rem;
  }
`;

export { CloseButton, Form, ModalButton };
