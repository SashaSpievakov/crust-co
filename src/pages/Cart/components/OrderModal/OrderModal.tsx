import { FC, useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';

import { Input, Modal } from '../../../../components/UI';
import { CloseButton, Form, ModalButton } from './OrderModal.styled';

interface OrderModalProps {
  setIsOpen: (value: boolean) => void;
}

interface FormValues {
  name: string;
  phone: string;
  city: string;
  address?: string;
}

export const OrderModal: FC<OrderModalProps> = ({ setIsOpen }) => {
  const [isSubmitted, setIsSubmitted] = useState(false);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({ mode: 'onChange' });

  const onSubmit: SubmitHandler<FormValues> = () => {
    setIsSubmitted(true);
  };

  const onClose = () => {
    setIsOpen(false);
  };

  return !isSubmitted ? (
    <Modal
      title="Order"
      subTitle="Please provide your information, and our operator will get in touch
          with you as soon as possible to finish your order"
      onClose={onClose}
    >
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Input
          id="inputName"
          label="Name"
          {...register('name', { required: true })}
          error={errors.name && 'This field is required'}
        />

        <Input
          id="inputPhone"
          label="Phone Number"
          type="number"
          {...register('phone', {
            required: {
              value: true,
              message: 'This field is required',
            },
            maxLength: {
              value: 10,
              message: 'The phone number should have 10 characters',
            },
            minLength: {
              value: 10,
              message: 'The phone number should have 10 characters',
            },
          })}
          error={errors.phone && errors.phone.message}
        />

        <Input
          id="inputCity"
          label="City"
          {...register('city', { required: true })}
          error={errors.city && 'This field is required'}
        />

        <Input
          id="inputAdddress"
          label="Adddress (optional)"
          {...register('address')}
        />

        <ModalButton type="submit">
          <span>Send</span>
        </ModalButton>
      </Form>
    </Modal>
  ) : (
    <Modal
      title="Thank you for your order!"
      subTitle="Our operator will give you a call in a few minutes"
      onClose={onClose}
    >
      <CloseButton onClick={onClose}>Close</CloseButton>
    </Modal>
  );
};
