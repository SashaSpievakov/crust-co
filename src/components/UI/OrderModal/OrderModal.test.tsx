import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CartItemsMockProps } from 'src/tests/mocks/mockData/mockData';
import OrderModal from './OrderModal';
import rendererWithAllProviders from '../../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../../tests/helpers/renderWithProviders';

describe('OrderModal Tests', () => {
  test('renders the OrderModal UI component', () => {
    const snapshot = rendererWithAllProviders(
      <OrderModal setIsOpen={jest.fn()} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('checks OrderModal logic', () => {
    beforeEach(() => {
      renderWithProviders(null, true, '/cart', {
        cart: CartItemsMockProps,
      });
    });

    test('checks OrderModal opening', () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      expect(
        screen.getByRole('heading', {
          name: 'Order',
        }),
      ).toBeInTheDocument();
    });

    test('closes OrderModal by clicking on the cross', () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      const cross = screen.getByTestId('modalCross');

      userEvent.click(cross);

      expect(
        screen.queryByRole('heading', {
          name: 'Order',
        }),
      ).not.toBeInTheDocument();
    });

    test('checks required fields error', async () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      const sumbitBtn = screen.getByRole('button', {
        name: /send/i,
      });

      userEvent.click(sumbitBtn);

      expect(
        await screen.findAllByText(/this field is required/i),
      ).toHaveLength(3);
    });

    test('checks min characters phone input error', async () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      const sumbitBtn = screen.getByRole('button', {
        name: /send/i,
      });
      const phoneInput = screen.getByRole('spinbutton');

      userEvent.type(phoneInput, '12353');
      userEvent.click(sumbitBtn);

      expect(
        await screen.findByText(/the phone number should have 10 characters/i),
      ).toBeInTheDocument();
    });

    test('checks max characters phone input error', async () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      const sumbitBtn = screen.getByRole('button', {
        name: /send/i,
      });
      const phoneInput = screen.getByRole('spinbutton');

      userEvent.type(phoneInput, '12345678901');
      userEvent.click(sumbitBtn);

      expect(
        await screen.findByText(/the phone number should have 10 characters/i),
      ).toBeInTheDocument();
    });

    test('successfully sumbits data', async () => {
      const buyButton = screen.getByRole('button', {
        name: /buy now/i,
      });

      userEvent.click(buyButton);

      const sumbitBtn = screen.getByRole('button', {
        name: /send/i,
      });
      const inputs = screen.getAllByRole('textbox');
      const phoneInput = screen.getByRole('spinbutton');

      userEvent.type(phoneInput, '4730395859');
      userEvent.type(inputs[0], 'Alex');
      userEvent.type(inputs[1], 'Toronto');
      userEvent.type(inputs[2], 'some address');
      userEvent.click(sumbitBtn);

      expect(
        await screen.findByText(/thank you for your order!/i),
      ).toBeInTheDocument();
    });
  });
});
