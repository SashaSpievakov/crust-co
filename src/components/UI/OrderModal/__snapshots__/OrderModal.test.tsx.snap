// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderModal Tests renders the OrderModal UI component 1`] = `
Object {
  "0": <div
    className="sc-dkrFOg ewVMXk"
    onClick={[Function]}
  />,
  "1": <article
    className="sc-bqWxrE boOcDi"
  >
    <h3
      className="sc-ksBlkl ljtipp"
    >
      Order
    </h3>
    <h4
      className="sc-hBxehG btPktv"
    >
      Please provide your information, and our operator will get in touch with you as soon as possible to finish your order
    </h4>
    <svg
      className="sc-fnGiBr hAjlmc"
      data-testid="modalCross"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      onKeyDown={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      tabIndex={0}
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1.293 1.293a1 1 0 0 1 1.414 0L8 6.586l5.293-5.293a1 1 0 1 1 1.414 1.414L9.414 8l5.293 5.293a1 1 0 0 1-1.414 1.414L8 9.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L6.586 8 1.293 2.707a1 1 0 0 1 0-1.414z"
      />
    </svg>
    <form
      className="sc-fEXmlR dWupFg"
      onSubmit={[Function]}
    >
      <div
        className="sc-bjfHbI iYBNeV"
      >
        <label
          className="sc-idXgbr eoKsmz"
          htmlFor="inputName"
        >
          Name
        </label>
        <input
          className="sc-dIfARi kdngOR"
          id="inputName"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </div>
      <div
        className="sc-bjfHbI iYBNeV"
      >
        <label
          className="sc-idXgbr eoKsmz"
          htmlFor="inputPhone"
        >
          Phone Number
        </label>
        <input
          className="sc-dIfARi kdngOR"
          id="inputPhone"
          name="phone"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyPress={[Function]}
          type="number"
        />
      </div>
      <div
        className="sc-bjfHbI iYBNeV"
      >
        <label
          className="sc-idXgbr eoKsmz"
          htmlFor="inputCity"
        >
          City
        </label>
        <input
          className="sc-dIfARi kdngOR"
          id="inputCity"
          name="city"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </div>
      <div
        className="sc-bjfHbI iYBNeV"
      >
        <label
          className="sc-idXgbr eoKsmz"
          htmlFor="inputAdddress"
        >
          Adddress (optional)
        </label>
        <input
          className="sc-dIfARi kdngOR"
          id="inputAdddress"
          name="address"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </div>
      <button
        className="sc-hLBbgP sc-gKPRtg sc-hHTYSt hEWbyi cwILTQ deIBzx"
        type="submit"
      >
        <span>
          Send
        </span>
      </button>
    </form>
  </article>,
}
`;
